status/barcode.py

```diff
@@ -0,0 +1,118 @@
+"""Handlers related to test for barcode printing
+"""
+from __future__ import print_function
+import subprocess
+import tornado.web
+import json
+from datetime import datetime
+from dateutil import parser
+from collections import Counter
+from status.util import SafeHandler
+import re
+
+class BarcodeHandler(SafeHandler):
+    """ Main script to print barcode and other plate labels on the barcode printer
+        Zebra located in the big lab at NGI
+        can distinguish between
+            - already existing barcode file
+            - text file
+            - text input
+    """
+    def get(self):
+        t = self.application.loader.load("barcode.html")
+        self.write(t.generate(gs_globals=self.application.gs_globals, user=self.get_current_user()))
+
+    def post(self):
+        # how often is the label(s) to be printed? Default: 1
+        copies = int(self.get_argument('copies'))
+
+        # printing from file input
+        try: # figure out if a file was entered in the form
+            file_for_printing = self.request.files['file_to_print']
+        except KeyError: # form entry for file was empty
+            print("no file was given")
+        else: #form entry for file was not empty
+            #retrieve file and transform the contained text to string format with utf-8 encoding
+            linesToPrint = str(file_for_printing[0]['body'], encoding='utf-8')
+            #check if file is already a label format file
+            barcodeFileMatch = re.compile('^\^XA')
+            isBarcodeFile = re.search(barcodeFileMatch, linesToPrint)
+            if isBarcodeFile: # if the file is already in the correct format it can be immediately moved to print
+                for _ in range(copies): # loops over copies to print
+                    print_barcode(linesToPrint)
+            else: # file submitted is a text file
+                array_linesToPrint = linesToPrint.splitlines() # split into the different lines of the text file
+                for line in array_linesToPrint:
+                    createdLabel = makeBarcode(line, match_barcode(line, False))
+                    createdLabel_joined = '\n'.join(createdLabel)
+                    for _ in range(copies): # loops over copies to print
+                        print_barcode(createdLabel_joined)
+
+        # printing from string input
+        text_for_printing = self.get_argument('text_to_print') # retrieves string from form
+        if len(text_for_printing) > 0 : # check wether there was an entry or not
+            createdLabel = makeBarcode(text_for_printing, match_barcode(text_for_printing, False))
+            createdLabel_joined = '\n'.join(createdLabel)
+            for _ in range(copies): # loops over copies to print
+                print_barcode(createdLabel_joined)
+        else:
+            print("no text was given")
+
+        self.set_status(200)
```

Comment:
Actually, this is not an error. This specifically is when the input was a file, then the text entry is empty. I just don't want to have an if statement without an else.

Reply:
@alneberg We would need javascript/jquery to actually handle the the submission to and response from the post method. The code currently gets around that by setting the target to a new tab. So it opens a new tab every time the form is submitted. It could be something to improve, as a next step maybe?

Reply:
Yeah, it can wait until next step I guess. But it used to be possible to have html forms with error handling without javascript no? Maybe I'm confused

Reply:
An ajax call would be the easiest way to handle it, I think. But we can do it as an enhancement.